Source/JavaScriptCore/runtime/WaiterListManager.cpp

```diff
@@ -50,38 +50,56 @@ WaiterListManager& WaiterListManager::singleton()
     return manager;
 }
 
-JSValue WaiterListManager::wait(JSGlobalObject* globalObject, VM& vm, void* ptr, AtomicsWaitValidation validation, Seconds timeout, AtomicsWaitType type)
+template <typename ValueType>
+JSValue WaiterListManager::waitImpl(JSGlobalObject* globalObject, VM& vm, ValueType* ptr, ValueType expectedValue, Seconds timeout, AtomicsWaitType type)
 {
     if (type == AtomicsWaitType::Async) {
         JSObject* object = constructEmptyObject(globalObject);
 
         bool isAsync = false;
         JSValue value;
-        if (validation == AtomicsWaitValidation::Fail)
-            value = vm.smallStrings.notEqualString();
-        else if (!timeout)
-            value = vm.smallStrings.timedOutString();
-        else {
-            isAsync = true;
-            JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
-            addAsyncWaiter(ptr, promise, timeout);
-            value = promise;
+
+        Ref<WaiterList> list = findOrCreateList(ptr);
+        JSPromise* promise = JSPromise::create(vm, globalObject->promiseStructure());
+        Ref<Waiter> waiter = adoptRef(*new Waiter(promise));
+
+        {
+            Locker listLocker { list->lock };
+            if (WTF::atomicLoad(ptr) != expectedValue)
+                value = vm.smallStrings.notEqualString();
+            else if (!timeout)
+                value = vm.smallStrings.timedOutString();
+            else {
+                isAsync = true;
+                list->addLast(listLocker, waiter);
+
+                if (timeout != Seconds::infinity()) {
+                    Ref<RunLoop::DispatchTimer> timer = RunLoop::current().dispatchAfter(timeout, [this, ptr, waiter = waiter.copyRef()]() mutable {
+                        timeoutAsyncWaiter(ptr, WTFMove(waiter));
+                    });
+                    waiter->setTimer(listLocker, WTFMove(timer));
+                }
+
+                value = promise;
+            }
         }
 
+        dataLogLnIf(WaiterListsManagerInternal::verbose && isAsync, "WaiterListManager added a new AsyncWaiter ", RawPointer(waiter.ptr()), " to a waiterList for ptr ", RawPointer(ptr));
```

Comment:
This looks not so great, in particular, `isAsync`.
You should only use `WaiterListsManagerInternal::verbose` for disable / enable of this dataLog.